src/linux.rs

```diff
@@ -30,6 +35,54 @@ pub fn get_mac(name: Option<&str>) -> Result<Option<[u8; 6]>, MacAddressError> {
     Ok(None)
 }
 
+/// Uses the `getifaddrs` call to retrieve a list of network interfaces on the
+/// host device and returns the MAC address that matching the adapter with
+/// the given IP
+///
+/// Because nix returns all of the IP's and MAC's in a combined list, we need
+/// to map the IP to an inteface name and track the MAC addresses by interface name
+/// and see if there's a match
+pub fn get_mac_address_by_ip(ip: &IpAddr) -> Result<Option<[u8; 6]>, MacAddressError> {
+    let ifiter = getifaddrs()?;
+
+    let mut ip_on_inferface: Option<String> = None;
+    let mut mac_to_interface: HashMap<String, Option<[u8; 6]>> = HashMap::new();
```

Comment:
I'm a little confused what the purpose of this `HashMap` is, if we already have the interface IP and MAC while we're iterating over them

Reply:
So, as I understand it, the ```getifaddrs()```'s call returns all of the addresses (MAC, IPv4, IPv6, etc.) of all of the interfaces on the system in a flat structure, e.g., there's no single record that has both the MAC and the IP address.  So there will (ideally) be one record with the target IP and a different record with the resulting MAC address.   The HashMap is just creating an index/map from each interface to its MAC address so, for example, if we find the MAC address entry first, we record it and then if we find the IP entry later, we can look it up and return it.  Similarly, because the order of the records that ```getifaddrs()``` returns isn't guaranteed, we could in theory find the IP address we're looking for first, and then the MAC address in a later record, so we have to handle that case as well.

Does that make sense?  Did I understand it correctly?  FYI: I did look in the nix crate for a more direct function to do the mapping but couldn't find it, so there may be a better way to do this.

Reply:
Thanks for the reviews!  FYI: I ran this on both a Linux and Windows system and "it worked for me" but it's wasn't clear how to do deeper testing for this type of tightly coupled system calls.  Let me know about the rest and I'll do another pass on the code.

Reply:
Hmm, I see. Thanks for the info, I'll do some further investigating and see what I can find wrt. that specific bit of code :thinking: